Add api backing for annotation verification

hudson-newey opened the pull request:
# Add api backing for annotation verification

This PR adds support for api/database backed verifications

## Changes

- Backfills verification api endpoints, services, and models
- Adds api backing for verification grid
- Bump web components version
- Makes AudioEvent models `channel` property writable
- Updates TypeScript typings to allow combination InnerFilter arrays
- Do not load source maps during testing to reduce the number of erroneous errors. If the source map doesn't exist in a third party library, the Karma browser console used to throw an error, making it hard to debug tests
- Verification grid speed is SIGNIFICANTLY FASTER to load

## Issues

Fixes: #2220 

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

hudson-newey committed:
Backfill verification services and tests

hudson-newey committed:
basic functionality

hudson-newey committed:
Handle failures

hudson-newey committed:
Add tests and dispose of decision model in verifications

hudson-newey committed:
Bump @ecoacoustics/web-components version

hudson-newey committed:
Update verification component to work with new format

hudson-newey committed:
bump web component version

hudson-newey committed:
Add support for overwriting verifications

hudson-newey committed:
Fix typing for filter combinations

hudson-newey committed:
Backfill verification tests

hudson-newey committed:
Add createOrUpdate tests

hudson-newey committed:
Full verification service testing

hudson-newey committed:
Fix verification service tests

hudson-newey committed:
Working verification page tests

hudson-newey committed:
Staging

hudson-newey committed:
Self review

hudson-newey committed:
Self review 2

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -96,6 +94,14 @@ class VerificationComponent
+    }
+
+    return User.getUnknownUser(undefined);
+  }
```
leaky concern - add the method to the session service?

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -218,6 +264,19 @@ class VerificationComponent
+    return (
+      event instanceof CustomEvent &&
+      event.detail instanceof Array &&
+      event.detail.length > 0
```
i mean, yeah, it's imperfect, but this is still quite loose. Perhaps we should check for at least one well known field on the first instance of the array? e.g. `subject`?

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -144,8 +150,48 @@ class VerificationComponent
+        false: ConfirmedStatus.Incorrect,
+        unsure: ConfirmedStatus.Unsure,
+        skip: ConfirmedStatus.Skip,
+      } as const satisfies Record<DecisionOptions, ConfirmedStatus>;
```
pull constant mapping out into field - don't redeclare in each invocation of the handler

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -144,8 +150,48 @@ class VerificationComponent
+      const mappedDecision =
+        confirmedMapping[subjectWrapper.verification.confirmed as string];
+
+      const tagId = subjectWrapper.tag.id;
```
have we shipped tags having `target`s  yet?

hudson-newey replied:
Not yet

https://github.com/ecoacoustics/web-components/blob/1919d1d2f83bf5e69bf9167c05470fa3738b3882/src/models/tag.ts

Above is a commit locked link to tag model

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -144,8 +150,48 @@ class VerificationComponent
+      const verification = new Verification(verificationData, this.injector);
+
+      // we need to subscribe otherwise the observable is never evaluated and
+      // the api request is never made
```
i've read toPromise resolves this?

hudson-newey replied:
`toPromise` is deprecated

see: https://stackoverflow.com/a/67044352

---

However, using `firstValueFrom` is the new equivalent.

![image](https://github.com/user-attachments/assets/94356b8f-bda5-4fef-98e6-4ccae1400872)

My second question would be why would `this.api.createOrUpdate.toPromise()` be better than `this.api.createOrUpdate.subscribe()`?

Using subscribe is more common in my mind, so it requires less explanation

atruskie replied:
because then you don't have to unsubscribe. You did the `takeUntil(this.unsubscribe)` dance so it's not a problem but to me, a promise reflects the once off intent better

hudson-newey replied:
I like the idea. I'm convinced

atruskie commented on src/app/models/Verification.ts:
```diff
@@ -0,0 +1,67 @@
+// I do not use the enum provided by the web components because the enum type
+// specified in the web components is a reflection of the web component
+// representation of the verification status, while this enum if a reflection of
+// the servers representation of the verification status
```
👍 yeah they are separate constructs and are not guaranteed to be related

atruskie commented on src/app/models/Verification.ts:
```diff
@@ -0,0 +1,67 @@
+  public audioEvent?: AudioEvent;
+
+  // TODO: Implement viewUrl once the verification view is implemented
+  // see: https://github.com/QutEcoacoustics/workbench-client/pull/2221
```
probably won't even have it's own view, if we're looking to fill this in for the sake of completion, it'd probably be the single annotation view route

atruskie commented on src/app/services/baw-api/verification/verification.service.ts:
```diff
@@ -0,0 +1,201 @@
+   * Creates a verification model if it doesn't already exist, if it already
+   * exists, update the existing model.
+   */
+  public createOrUpdate(
```
todo: simplify when baw-server implements an upsert route
also file an issue on baw-server

hudson-newey replied:
Associated issue: https://github.com/QutEcoacoustics/baw-server/issues/724

atruskie commented on src/app/services/baw-api/verification/verification.service.ts:
```diff
@@ -0,0 +1,201 @@
+      map((results) => (results.length > 0 ? results[0] : null))
+    );
+  }
+}
```
i mean you did your best, but oof, having to do a filter here sucks

you can limit the page size to 1 thought - that will help

hudson-newey replied:
Updated code
![image](https://github.com/user-attachments/assets/f2e173b5-168c-4ebd-a60c-370684110126)

atruskie reviewed (approved):
I like how the verification component is not that much more complex. Good job bundling the logic up into the service.

hudson-newey committed:
Review feedback

hudson-newey committed:
Use firstValueFrom instead of subscribe

hudson-newey committed:
Use firstValueFrom instead of subscribe

hudson-newey committed:
Fix tests

hudson-newey committed:
Fix spectrograms with correct "stretch" rules

hudson-newey committed:
Fix verification service tests

hudson-newey committed:
Push unstaged fixes for new web component version

I forgot to push these changes from my work computer last week

hudson-newey committed:
Bump web component versions

hudson-newey committed:
Fix verification.component.spec tag typeahead mocks

hudson-newey committed:
Partial hacky fixes for verification component

This commit fixes the tests for the api backed verification component.
However, it does not assert what createOrUpdate is called with (hence
the hacky comment)

A following commit will be added to assert that the createOrUpdate
service method is called with the correct decision content and clean up
a lot of the hacky code.

This commit is being pushed to test in CI

hudson-newey committed:
Use consistent window size for verification tests

These tests were previously flakey because the verification grid size
would change depending on the Karma browser window.

This commit sets all verification component tests to the same size

hudson-newey committed:
Fix for getPage callback updates

This commit is hacky and I should find the root cause of why I need this
hack

hudson-newey committed:
Revert web component bump

Bump was faulty because of cached deps

hudson-newey committed:
Re-bump web-components version

hudson-newey committed:
Hacky fix for grid tile content

hudson-newey committed:
Working CI?

hudson-newey committed:
Resiliance hacks

hudson-newey committed:
Natural scaling spectrograms and better inline tags

hudson-newey committed:
Fix tests for annotation event card

hudson-newey committed:
Revert natural scaling
